### PR TITLE
HPCC-17569 Add roxie support for persistent http connections

### PR DESCRIPTION
**Pull request by afishbeck**
Signed-off-by: Anthony Fishbeck <anthony.fishbeck@lexisnexis.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by afishbeck** [comment]
@mayx please review

**Comment by mayx** [comment]
@richardkchapman @afishbeck  sorry I must've missed the notice email. Will get on it immediately.

File: common/thorhelper/roxiehelper.cpp [hunk 1]
```diff
@@ -1771,6 +1780,8 @@ class HttpResponseHandler
             compression = respCompression;
         header.append("HTTP/1.0 200 OK\r\n");
         header.append("Content-Type: ").append(mlFmt == MarkupFmt_JSON ? "application/json" : "text/xml").append("\r\n");
+        if (httpKeepAlive)
+            header.append("Connection: Keep-alive\r\n");
```
Comment (mayx) [inline common/thorhelper/roxiehelper.cpp]: Keep-Alive

File: common/thorhelper/roxiehelper.hpp [hunk 2]
```diff
@@ -65,6 +66,13 @@ class THORHELPER_API HttpHelper : public CInterface
     HttpHelper(StringArray *_validTargets) : validTargets(_validTargets), method(HttpMethod::NONE) {parameters.setown(createProperties(true));}
     inline bool isHttp() { return method!=HttpMethod::NONE; }
     inline bool isHttpGet(){ return method==HttpMethod::GET; }
+    inline bool allowKeepAlive()
+    {
+        const char *connection = queryRequestHeader("Connection");
+        if (!connection)
+            return streq(version, "1.1");
```
Comment (mayx) [inline common/thorhelper/roxiehelper.hpp]: Would !streq(version, "1.0") be better, in which case the code doesn't have to be changed for newer http versions.

File: common/thorhelper/roxiehelper.hpp [hunk 3]
```diff
@@ -65,6 +66,13 @@ class THORHELPER_API HttpHelper : public CInterface
...
+        return strieq(connection, "Keep-alive");
```
Comment (mayx) [inline common/thorhelper/roxiehelper.hpp]: Keep-Alive

File: common/thorhelper/roxiehelper.cpp (hunk 1 above)
Comment (mayx) [reply common/thorhelper/roxiehelper.cpp]: Do we still want to use HTTP/1.0, or should we switch it to HTTP/1.1, and if httpKeepAlive is false, set "Connection: Close" ?

**Review by mayx — COMMENTED** [review]
@afishbeck a few minor comments

File: common/thorhelper/roxiehelper.cpp (hunk 1 above)
Comment (afishbeck) [reply common/thorhelper/roxiehelper.cpp]: I didn't want to chance any interop issues just for this issue.  We can evaluate changing to HTTP/1.1 separately.

**Comment by afishbeck** [comment]
@mayx updated (and squashed just to save time)

**Review by mayx — APPROVED** [review]
@afishbeck Looking good now.

**Pull request merged**